File: sdk/python/tests/e2e_query.py
"""E2E: query surface — custom ids, states, counts, logs, progress, children."""

from __future__ import annotations

import time

from harness import Server, test, unique_name

from bunqueue import CommandError, Queue


@test
def custom_id_idempotent(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        first = queue.add("task", {"v": 1}, job_id="my-custom-id")
        second = queue.add("task", {"v": 2}, job_id="my-custom-id")
        assert first.id == second.id  # idempotent re-add returns same job
        fetched = queue.get_job_by_custom_id("my-custom-id")
        assert fetched is not None and fetched.id == first.id


@test
def get_jobs_by_state(server: Server) -> None:
    from harness import wait_until

    with Queue(unique_name("q"), port=server.port) as queue:
        queue.add_bulk([{"name": "t", "data": {"i": i}} for i in range(5)])
        queue.add("later", {}, delay=60000)
        # GetJobs reads persisted rows: allow the 10ms write buffer to flush
        assert wait_until(lambda: len(queue.get_waiting()) == 5, 10), queue.get_waiting()
        assert wait_until(lambda: len(queue.get_delayed()) == 1, 10)
        assert queue.get_waiting_count() == 5
        assert queue.get_delayed_count() == 1
        both = queue.get_jobs(["waiting", "delayed"])
        assert len(both) == 6, f"multi-state returned {len(both)}"


@test
def counts_per_priority(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        queue.add("a", {}, priority=1)
        queue.add("b", {}, priority=1)
        queue.add("c", {}, priority=5)
        counts = queue.get_counts_per_priority()
        assert counts, f"empty counts: {counts}"
        total = sum(int(v) for v in counts.values())
        assert total == 3


@test
def progress_roundtrip(server: Server) -> None:
    """Progress requires the job to be active: pull it first."""
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {})
        queue.connection.call({"cmd": "PULL", "queue": queue.name, "owner": "e2e", "timeout": 1000})
        queue.update_job_progress(job.id, 42, "almost")
        progress = queue.get_progress(job.id)
        assert progress["progress"] == 42
        assert progress["message"] == "almost"


@test
def job_logs(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {})
        queue.add_job_log(job.id, "step one")
        queue.add_job_log(job.id, "step two")
        logs = queue.get_job_logs(job.id)
        assert len(logs) == 2, f"expected 2 logs, got {logs}"
        queue.clear_job_logs(job.id)
        assert len(queue.get_job_logs(job.id)) == 0


@test
def tags_and_group_roundtrip(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {"x": 1}, tags=["alpha", "beta"], group_id="g1", priority=7)
        fetched = queue.get_job(job.id)
        assert fetched is not None
        assert fetched.tags == ["alpha", "beta"]
        assert fetched.group_id == "g1"
        assert fetched.priority == 7
        assert fetched.name == "t"


@test
def wait_for_job_returns_result(server: Server) -> None:
    from bunqueue import Worker

    with Queue(unique_name("q"), port=server.port) as queue:
        worker = Worker(queue.name, lambda j: {"answer": 42}, port=server.port, poll_timeout_ms=500)
        worker.start()
        try:
            job = queue.add("t", {})
            result = queue.wait_for_job(job.id, timeout_ms=10000)
            assert result == {"answer": 42}, f"got {result}"
            job2 = queue.add("t2", {})
            assert queue.wait_job_until_finished(job2.id, ttl_ms=10000) == {"answer": 42}
        finally:
            worker.close(timeout=10)


@test
def nonexistent_job_state(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        try:
            state = queue.get_state("no-such-id")
            assert state in ("unknown", "None", "not_found"), f"unexpected state {state}"
        except CommandError:
            pass  # error response is acceptable too

File: sdk/typescript/tests/e2e-query.ts
/** E2E: basics + query surface — custom ids, states, counts, logs, progress. */

import { CommandError, Connection } from '../dist/legacy.js';
import { assert, assertEq, getPort, makeQueue, makeWorker, test, waitFor } from './harness.ts';

test('ping + hello', async () => {
  const conn = new Connection({ host: '127.0.0.1', port: getPort() });
  assert(await conn.ping(), 'ping should return true');
  const hello = await conn.hello();
  assert(Number(hello.protocolVersion) >= 1, 'protocolVersion >= 1');
  assert(typeof hello.server === 'string', 'server name present');
  conn.close();
});

test('add + getJob + name + state', async () => {
  const queue = makeQueue<{ to: string }>('add');
  try {
    const job = await queue.add('send', { to: 'a@b.c' }, { priority: 5 });
    assert(job.id.length > 0, 'job id assigned');
    const state = await queue.getJobState(job.id);
    assert(state === 'waiting' || state === 'prioritized', `state waiting/prioritized (${state})`);
    const fetched = await queue.getJob(job.id);
    assert(fetched !== null, 'job fetched');
    assertEq(fetched!.name, 'send', 'job name');
    assertEq((fetched!.data as Record<string, unknown>).to, 'a@b.c', 'job data');
  } finally {
    queue.close();
  }
});

test('addBulk 50 + count', async () => {
  const queue = makeQueue('bulk');
  try {
    const jobs = await queue.addBulk(
      Array.from({ length: 50 }, (_, i) => ({ name: 'task', data: { i } }))
    );
    assertEq(jobs.length, 50, 'bulk returned 50 jobs');
    assertEq(await queue.count(), 50, 'count is 50');
  } finally {
    queue.close();
  }
});

test('query: custom id idempotent + getJobByCustomId', async () => {
  const queue = makeQueue('customid');
  try {
    const first = await queue.add('task', { v: 1 }, { jobId: 'my-custom-id' });
    const second = await queue.add('task', { v: 2 }, { jobId: 'my-custom-id' });
    assertEq(second.id, first.id, 'idempotent re-add returns same id');
    const fetched = await queue.getJobByCustomId('my-custom-id');
    assert(fetched !== null && fetched.id === first.id, 'fetched by custom id');
    const dedupId = await queue.getDeduplicationJobId('my-custom-id');
    assertEq(dedupId, first.id, 'getDeduplicationJobId');
  } finally {
    queue.close();
  }
});

test('query: getJobs by state (multi-state, write-buffer flush)', async () => {
  const queue = makeQueue('states');
  try {
    await queue.addBulk(Array.from({ length: 5 }, (_, i) => ({ name: 't', data: { i } })));
    await queue.add('later', {}, { delay: 60_000 });
    // GetJobs reads persisted rows: allow the 10ms write buffer to flush
    await waitFor(async () => (await queue.getWaiting()).length === 5, 10_000);
    await waitFor(async () => (await queue.getDelayed()).length === 1, 10_000);
    assertEq(await queue.getWaitingCount(), 5, 'waiting count');
    assertEq(await queue.getDelayedCount(), 1, 'delayed count');
    const both = await queue.getJobs({ state: ['waiting', 'delayed'] });
    assertEq(both.length, 6, 'multi-state getJobs');
  } finally {
    queue.close();
  }
});

test('query: counts per priority', async () => {
  const queue = makeQueue('perprio');
  try {
    await queue.add('a', {}, { priority: 1 });
    await queue.add('b', {}, { priority: 1 });
    await queue.add('c', {}, { priority: 5 });
    const counts = await queue.getCountsPerPriority();
    const total = Object.values(counts).reduce((a, b) => a + Number(b), 0);
    assertEq(total, 3, 'per-priority totals add up');
  } finally {
    queue.close();
  }
});

test('query: progress roundtrip (job must be active)', async () => {
  const queue = makeQueue('progress');
  try {
    const job = await queue.add('t', {});
    // Progress requires the job to be ACTIVE: pull it first
    await queue.connection.call({ cmd: 'PULL', queue: queue.name, owner: 'e2e', timeout: 1000 });
    await queue.updateJobProgress(job.id, 42, 'almost');
    const progress = await queue.getProgress(job.id);
    assertEq(progress.progress, 42, 'progress value');
    assertEq(progress.message, 'almost', 'progress message');
  } finally {
    queue.close();
  }
});

test('query: job logs add / get / clear', async () => {
  const queue = makeQueue('logs');
  try {
    const job = await queue.add('t', {});
    await queue.addJobLog(job.id, 'step one');
    await queue.addJobLog(job.id, 'step two');
    const logs = await queue.getJobLogs(job.id);
    assertEq(logs.length, 2, 'two log rows');
    assert(logs[0].includes('step one'), 'first log row content');
    await queue.clearJobLogs(job.id);
    assertEq((await queue.getJobLogs(job.id)).length, 0, 'logs cleared');
  } finally {
    queue.close();
  }
});

test('query: tags + groupId + priority roundtrip', async () => {
  const queue = makeQueue('tags');
  try {
    const job = await queue.add(
      't',
      { x: 1 },
      { tags: ['alpha', 'beta'], groupId: 'g1', priority: 7 }
    );
    const fetched = await queue.getJob(job.id);
    assert(fetched !== null, 'job fetched');
    assertEq(fetched?.tags.join(','), 'alpha,beta', 'tags roundtrip');
    assertEq(fetched?.groupId, 'g1', 'groupId roundtrip');
    assertEq(fetched?.priority, 7, 'priority roundtrip');
    assertEq(fetched?.name, 't', 'name roundtrip');
  } finally {
    queue.close();
  }
});

test('query: waitForJob returns worker result', async () => {
  const queue = makeQueue('waitjob');
  const worker = makeWorker(queue.name, async () => ({ answer: 42 }));
  try {
    const job = await queue.add('t', {});
    const result = await queue.waitForJob<{ answer: number }>(job.id, 10_000);
    assertEq(result.answer, 42, 'waitForJob result');
  } finally {
    await worker.close();
    queue.close();
  }
});

test('query: nonexistent job → getJob null, getJobState tolerated', async () => {
  const queue = makeQueue('missing');
  try {
    assertEq(await queue.getJob('no-such-id'), null, 'getJob returns null');
    try {
      const state = await queue.getJobState('no-such-id');
      assert(['unknown', 'not_found'].includes(state), `unexpected state ${state}`);
    } catch (err) {
      assert(err instanceof CommandError, 'CommandError acceptable');
    }
  } finally {
    queue.close();
  }
});
